In [1]:
import math
from typing import Callable
import numpy as np

def est_D(probabilidades: list[float]):
    n = len(probabilidades)
    prob = sorted(probabilidades)

    D_p = [((i + 1)/n - prob[i]) for i in range(n)]  # Fe(Y) - F(Y)
    D_m = [(prob[i] - i/n) for i in range(n)]  # F(Y) - Fe(Y)

    d = max(max(D_p), max(D_m))

    return d

def estimar_p_valor_simulaciones_continua(
    d_obs: float,
    gen_prob: Callable[[], list[float]],
    N_SIM: int = 10_000
):
    # Se utiliza el estadístico de Kolmogorov-Smirnov

    p_valor = 0.0
    for _ in range(N_SIM):
        probabilidades = gen_prob()
        d_j = est_D(probabilidades)

        if d_j >= d_obs:
            p_valor += 1

    p_valor = p_valor / N_SIM
    return p_valor

### Ejercicio 7.
Generar los valores correspondientes a $30$ variables aleatorias exponenciales independientes, cada una con media $1$. Luego, en base al estadístico de prueba de Kolmogorov-Smirnov, aproxime el *p−valor* de la prueba de que los datos realmente provienen de una distribución exponencial con media $1$.

---

### Análisis
$$
\begin{array}{cl}

H_0 & = \text{"Los 30 datos provienen de una distribución exponencial con media 1"} \\
H_1 & = \text{"Los 30 datos no provienen de una distribución exponencial con media 1"}

\end{array}
$$

Asumo $\alpha = 0.1$

In [3]:
n = 30
muestra = [(-math.log(1 - np.random.random())) for _ in range(n)]  # lambda = 1

def F_H0(x: float):
    # lambda = 1
    return 1 - math.exp(-x)

probabilidades = [F_H0(muestra[i]) for i in range(n)]

d_obs = est_D(probabilidades)
p_valor = estimar_p_valor_simulaciones_continua(
    d_obs,
    lambda: np.random.uniform(0, 1, n) # Como no se estima ningún parámetro, se puede utilizar la uniforme
)

print("d_obs: ", d_obs)
print("p_valor: ", p_valor)

d_obs:  0.13582895098196035
p_valor:  0.5856


Como $\text{p-valor} \gt \alpha \equiv 0.5856 \gt 0.1$, entonces no hay suficiente evidencia para rechazar $H_0$. 